geonode/security/models.py

```diff
@@ -137,16 +129,24 @@ def set_default_permissions(self):
                         anonymous_group, self.get_self_resource())
 
         if self.__class__.__name__ == 'Layer':
-            if anonymous_can_view and settings.OGC_SERVER['default'].get(
-                    "GEOFENCE_SECURITY_ENABLED", False):
-                if not getattr(settings, 'DELAYED_SECURITY_SIGNALS', False):
-                    set_geofence_all(self)
-                else:
-                    self.get_self_resource().set_dirty_state()
             # only for layer owner
             assign_perm('change_layer_data', self.owner, self)
             assign_perm('change_layer_style', self.owner, self)
-
+            if settings.OGC_SERVER['default'].get("GEOFENCE_SECURITY_ENABLED", False):
+                perms = [
+                    "change_resourcebase",
+                    "change_resourcebase_permissions",
+                    "view_resourcebase",
+                    "download_resourcebase"]
+                if not getattr(settings, 'DELAYED_SECURITY_SIGNALS', False):
+                    if anonymous_can_view:
+                        set_geofence_all(self)
+                    else:
+                        sync_geofence_with_guardian(self.layer, perms, user=self.owner)
+                    if settings.DEFAULT_ANONYMOUS_DOWNLOAD_PERMISSION:
+                        sync_geofence_with_guardian(self.layer, perms, group=anonymous_group)
```

Comment:
@afabiani imho anonymous shouldn't have change_resourcebase and change_resourcebase_permissions assigned.

Reply:
@giohappy right, my fault. Thanks for pointing out.